NotRetryableError added

yarsky-tgz opened the pull request.

Commit:
Added NotRetryableError to add ability to break the retry process

Commit:
README updated

yarsky-tgz commented:
I have added test for added error, but coverage check fails.

normartin commented:
> I have added test for added error, but coverage check fails.

That is really strange. I added `process.exit();` statement to the if branch and the test still passes. We need to fix this first.

normartin commented on test/notretryable.error.test.ts:
```diff
@@ -0,0 +1,24 @@
+import {expect} from "./index";
+import {customizeRetry, defaultRetryConfig, notEmpty, retry, wait, NotRetryableError, RetryError} from "../src/retry-promise";
```
Can you cleanup unused imports?

normartin commented on src/retry-promise.ts:
```diff
@@ -96,6 +96,10 @@ async function _retry<T>(f: () => Promise<T>, config: RetryConfig<T>, done: () =
         } catch (error) {
+            if (error.name === NotRetryableError.name) throw new RetryError(
+              `Met not retryable error. Last error: ${lastError!}`,
+              lastError!
```
I think we do not need, not null assertions in both cases.

yarsky-tgz replied:
yeah, and wrong variable was used, please check, hope resolved

yarsky-tgz commented:
> > I have added test for added error, but coverage check fails.
> 
> That is really strange. I added `process.exit();` statement to the if branch and the test still passes. We need to fix this first.

can you show please, where exactly you've added `process.exit();` ?

normartin reviewed (commented):
> > > I have added test for added error, but coverage check fails.
> > 
> > 
> > That is really strange. I added `process.exit();` statement to the if branch and the test still passes. We need to fix this first.
> 
> can you show please, where exactly you've added `process.exit();` ?



> > > I have added test for added error, but coverage check fails.
> > 
> > 
> > That is really strange. I added `process.exit();` statement to the if branch and the test still passes. We need to fix this first.
> 
> can you show please, where exactly you've added `process.exit();` ?

I tried on my machine. 
```
if (error.name === NotRetryableError.name){
  process.exit();
   throw new RetryError(
              `Met not retryable error. Last error: ${lastError!}`,
              lastError!`);
}
```
This should make the test fail. But it does not. I have no idea why.

Commit:
removed redundant null assertions, fixed error variable name

Commit:
cleared unused imports

yarsky-tgz commented:
![image](https://user-images.githubusercontent.com/11916002/96909055-27376300-14a6-11eb-9373-07b9dee1f5b9.png)
![image](https://user-images.githubusercontent.com/11916002/96909136-47ffb880-14a6-11eb-8ec5-4a2ea95de96f.png)

yarsky-tgz commented:
```
console.log((new NotRetryableError()).constructor.name)
```

```
Error
```


![luv-films-1gEbYgyzAN9Az2oywK](https://media.giphy.com/media/l0IyjxxfGfdqEiVfa/giphy-downsized.gif)

Commit:
wrapped error and defined name for it

https://stackoverflow.com/questions/31626231/custom-error-class-in-typescript

yarsky-tgz commented:
found solution here:

https://stackoverflow.com/questions/31626231/custom-error-class-in-typescript

yarsky-tgz commented:
Do you accept idea and implementation as it looks now?

normartin commented:
Yes, looks good! Will release new version now.
